snyk/snyk-nuget-plugin
# feat: [OSM-516] Add runtime assembly constraints to depGraph

This PR will be split up in smaller parts, it became way too big.

# PR Before the close


#### What does this PR do?

Picks up where https://github.com/snyk/snyk-nuget-plugin/pull/147 left off.

Add support for overriding the dependency graph's runtime dependency versions with the assembly versions detected upon doing a `dotnet publish` for newer dotnet core projects.

#### Where should the reviewer start?
Go from `inspect()` and see the path you take when scanning a dotnet core project with `dotnet-runtime-resolution` in the `options`.

#### How should this be manually tested?
* Check out the CLI
* `npm link` this version to it
* Go to your Snyk account and activate the feature flag `useIsolatedBuildsForDotnet` (yes, that name makes no sense in this context)
* Run `snyk test` through the CLI version you did an npm link with

#### What are the relevant tickets?
https://snyksec.atlassian.net/browse/OSM-515

#### Screenshots

For the following `.csproj`

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <RuntimeIdentifier>osx-x64</RuntimeIdentifier>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="NSubstitute" Version="4.3.0" />
  </ItemGroup>
</Project>
```

Before:

```
snyk test --print-deps
IfDotNoIssue @ 1.0.0
   └─ NSubstitute @ 4.3.0
      └─ Castle.Core @ 4.4.1
         ├─ NETStandard.Library @ 1.6.1
         │  ├─ Microsoft.NETCore.Platforms @ 1.1.0
         │  ├─ Microsoft.Win32.Primitives @ 4.3.0
         │  │  ├─ Microsoft.NETCore.Platforms @ 1.1.0
         │  │  ├─ Microsoft.NETCore.Targets @ 1.1.0
         │  │  └─ System.Runtime @ 4.3.0
         │  │     ├─ Microsoft.NETCore.Platforms @ 1.1.0
         │  │     └─ Microsoft.NETCore.Targets @ 1.1.0
         │  ├─ System.AppContext @ 4.3.0
         │  │  └─ System.Runtime @ 4.3.0
         │  ├─ System.Collections @ 4.3.0
         │  │  ├─ Microsoft.NETCore.Platforms @ 1.1.0
         │  │  ├─ Microsoft.NETCore.Targets @ 1.1.0
         │  │  └─ System.Runtime @ 4.3.0
         │  ├─ System.Collections.Concurrent @ 4.3.0
         │  │  ├─ System.Collections @ 4.3.0
         │  │  ├─ System.Diagnostics.Debug @ 4.3.0
         │  │  │  ├─ Microsoft.NETCore.Platforms @ 1.1.0
         │  │  │  ├─ Microsoft.NETCore.Targets @ 1.1.0
(...)
Tested 72 dependencies for known issues, found 6 issues, 7 vulnerable paths.


Issues with no direct upgrade or patch:
  ✗ Denial of Service (DoS) [High Severity][https://security.snyk.io/vuln/SNYK-DOTNET-SYSTEMNETHTTP-60045] in System.Net.Http@4.3.0
    introduced by NSubstitute@4.3.0 > Castle.Core@4.4.1 > NETStandard.Library@1.6.1 > System.Net.Http@4.3.0
  This issue was fixed in versions: 4.1.2, 4.3.2
  ✗ Improper Certificate Validation [High Severity][https://security.snyk.io/vuln/SNYK-DOTNET-SYSTEMNETHTTP-60046] in System.Net.Http@4.3.0
    introduced by NSubstitute@4.3.0 > Castle.Core@4.4.1 > NETStandard.Library@1.6.1 > System.Net.Http@4.3.0
  This issue was fixed in versions: 4.1.2, 4.3.2
  ✗ Privilege Escalation [High Severity][https://security.snyk.io/vuln/SNYK-DOTNET-SYSTEMNETHTTP-60047] in System.Net.Http@4.3.0
    introduced by NSubstitute@4.3.0 > Castle.Core@4.4.1 > NETStandard.Library@1.6.1 > System.Net.Http@4.3.0
  This issue was fixed in versions: 4.1.2, 4.3.2
  ✗ Authentication Bypass [Medium Severity][https://security.snyk.io/vuln/SNYK-DOTNET-SYSTEMNETHTTP-60048] in System.Net.Http@4.3.0
    introduced by NSubstitute@4.3.0 > Castle.Core@4.4.1 > NETStandard.Library@1.6.1 > System.Net.Http@4.3.0
  This issue was fixed in versions: 4.1.2, 4.3.2
  ✗ Information Exposure [High Severity][https://security.snyk.io/vuln/SNYK-DOTNET-SYSTEMNETHTTP-72439] in System.Net.Http@4.3.0
    introduced by NSubstitute@4.3.0 > Castle.Core@4.4.1 > NETStandard.Library@1.6.1 > System.Net.Http@4.3.0
  This issue was fixed in versions: 2.0.20710, 4.0.1-beta-23225, 4.1.4, 4.3.4
  ✗ Regular Expression Denial of Service (ReDoS) [High Severity][https://security.snyk.io/vuln/SNYK-DOTNET-SYSTEMTEXTREGULAREXPRESSIONS-174708] in System.Text.RegularExpressions@4.3.0
    introduced by NSubstitute@4.3.0 > Castle.Core@4.4.1 > NETStandard.Library@1.6.1 > System.Text.RegularExpressions@4.3.0 and 1 other path(s)
  This issue was fixed in versions: 4.3.1
```

After:

```
snyk test --dotnet-runtime-resolution --print-deps
IfDotNoIssue @ 1.0.0
   └─ NSubstitute @ 4.3.0
      └─ Castle.Core @ 4.4.1
         ├─ NETStandard.Library @ 1.6.1
         │  ├─ Microsoft.NETCore.Platforms @ 1.1.0
         │  ├─ Microsoft.Win32.Primitives @ 6.0.0
         │  ├─ System.AppContext @ 6.0.0
         │  ├─ System.Collections @ 6.0.0
         │  ├─ System.Collections.Concurrent @ 6.0.0
         │  ├─ System.Console @ 6.0.0
         │  ├─ System.Diagnostics.Debug @ 6.0.0
         │  ├─ System.Diagnostics.Tools @ 6.0.0
         │  ├─ System.Diagnostics.Tracing @ 6.0.0
         │  ├─ System.Globalization @ 6.0.0
(...)

✔ Tested 55 dependencies for known issues, no vulnerable paths found.
```
